### PR TITLE
Feature - ThienLuong - TableView nang cao 

### PR DESCRIPTION
**Pull request by at-thienluong**
TableView nang cao

### CODE REVIEW

File: BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/Custom Gender Picker/GenderPicker.swift [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    
+    // MARK: - Properties
+    var gender = ["Male", "Female", "Not prefered"]
+    var selected = ""
+    
+    weak var delegate: GenderPickerDelegate?
+    
+    // MARK: - Life cycle
+    override func awakeFromNib() {
+        genderPicker.delegate = self
```
Comment (at-bienle) [inline BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/Custom Gender Picker/GenderPicker.swift]: Add super

File: BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/Custom Gender Picker/GenderPicker.swift [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        genderPicker.dataSource = self
+    }
+    
+    // MARK: - Action
+    @IBAction func doneButtonTouchUpInside(_ sender: Any) {
```
Comment (at-bienle) [inline BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/Custom Gender Picker/GenderPicker.swift]: Please check it in all your code

File: BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/Custom Gender Picker/GenderPicker.swift [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        delegate?.view(_view: self, needperformAction: .getGender(gender: selected))
+    }
+    
+    @IBAction func cancelButtonTouchUpInside(_ sender: Any) {
+        return
```
Comment (at-bienle) [inline BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/Custom Gender Picker/GenderPicker.swift]: If this function do nothing, no need return

File: BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/EditContactViewController/ContactEditViewController.swift [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+        
+        let alert = UIAlertController(title: title, message: message, preferredStyle: UIAlertController.Style.alert)
+        alert.addAction(UIAlertAction(title: "OK", style: UIAlertAction.Style.default, handler: nil))
+        present(alert, animated: true, completion: nil)
+    }
+    
+    // MARK: - Date Picker
+    private func showDatePicker(){
+        datePicker.datePickerMode = .date
+        let toolbar = UIToolbar();
```
Comment (at-bienle) [inline BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/EditContactViewController/ContactEditViewController.swift]: Please check `;` in this function

File: BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/EditContactViewController/ContactEditViewController.swift [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+        toolbar.sizeToFit()
+        let doneButton = UIBarButtonItem(title: "Done", style: .plain, target: self, action: #selector(doneDatePicker));
+        let spaceButton = UIBarButtonItem(barButtonSystemItem: UIBarButtonItem.SystemItem.flexibleSpace, target: nil, action: nil)
+        let cancelButton = UIBarButtonItem(title: "Cancel", style: .plain, target: self, action: #selector(cancelDatePicker));
+        toolbar.setItems([doneButton,spaceButton,cancelButton], animated: false)
+        dateTextField.inputAccessoryView = toolbar
+        dateTextField.inputView = datePicker
+    }
+    
+    @objc private func doneDatePicker(){
```
Comment (at-bienle) [inline BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/EditContactViewController/ContactEditViewController.swift]: Add space after `()`
Please check it in two function below

File: BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/EditContactViewController/ContactEditViewController.swift [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+        let formatter = DateFormatter()
+        formatter.dateFormat = "dd/MM/yyyy"
+        dateTextField.text = formatter.string(from: datePicker.date)
+        self.view.endEditing(true)
```
Comment (at-bienle) [inline BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/EditContactViewController/ContactEditViewController.swift]: Remove self, please check it in all your codes

File: BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/EditContactViewController/ContactEditViewController.swift [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+    }
+    
+    @objc private func cancelDatePicker(){
+        self.view.endEditing(true)
+    }
+    
+    // MARK: - Gender picker
+    @IBAction private func genderTextFieldEditing(_ sender: Any) {
+        guard let nib = Bundle.main.loadNibNamed("GenderPicker", owner: self, options: nil)! [0] as? GenderPicker else { return }
```
Comment (at-bienle) [inline BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/EditContactViewController/ContactEditViewController.swift]: Please check this line

File: BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/MasterContactViewController/ContactMasterViewController.swift [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+    // MARK: - Private func
+    private func makeNaviButton() {
+        let multiSelect = UIBarButtonItem(barButtonSystemItem: .trash, target: self, action: #selector(turnOnMultiSelect))
+        navigationItem.leftBarButtonItem = multiSelect
+        let addButton = UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(addButtonTouchUpInside))
+        navigationItem.rightBarButtonItem = addButton
+    }
+    
+    @objc private func turnOnMultiSelect() {
+        if tableView.isEditing == true { return } else {
```
Comment (at-bienle) [inline BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/MasterContactViewController/ContactMasterViewController.swift]: Please check this line

File: BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/MasterContactViewController/ContactMasterViewController.swift [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+// MARK: - UITableViewDataSource
+extension ContactMasterViewController: UITableViewDataSource {
+    
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return contactList.count
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: indentify, for: indexPath) as? ContactCustomTableViewCell else { return UITableViewCell() }
+        if let date = contactList[indexPath.row]["date"]?.suffix(4) {
```
Comment (at-bienle) [inline BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/MasterContactViewController/ContactMasterViewController.swift]: Please remove `?`, use unwrap

File: BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/MasterContactViewController/ContactMasterViewController.swift [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+extension ContactMasterViewController: UITableViewDelegate {
+    
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        if tableView.isEditing == false {
+            let contactEditViewController = ContactEditViewController()
+            contactEditViewController.contact = contacts[indexPath.row]
+            contactEditViewController.indexPath = indexPath
+            contactEditViewController.delegate = self
+            navigationController?.pushViewController(contactEditViewController, animated: true)
+            tableView.deselectRow(at: indexPath, animated: true)
```
Comment (at-bienle) [inline BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/MasterContactViewController/ContactMasterViewController.swift]: Move `deselect` to top

File: BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/MasterContactViewController/ContactMasterViewController.swift [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+        }
+        return cell
+    }
+}
+
+// MARK: - UITableViewDelegate
+extension ContactMasterViewController: UITableViewDelegate {
+    
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        if tableView.isEditing == false {
```
Comment (at-bienle) [inline BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/MasterContactViewController/ContactMasterViewController.swift]: No need `== false`

File: BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/NewContactViewController/ContactNewViewController.swift [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+            let imagePickerController = UIImagePickerController()
+            imagePickerController.delegate = self
+            imagePickerController.sourceType = sourceType
+            present(imagePickerController, animated: true, completion: nil)
+        }
+    }
+    
+    // MARK: - Action
+    @IBAction func genderTextFieldEdit(_ sender: Any) {
+        guard let nib = Bundle.main.loadNibNamed("GenderPicker", owner: self, options: nil)! [0] as? GenderPicker else { return }
```
Comment (at-bienle) [inline BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/NewContactViewController/ContactNewViewController.swift]: Please check this line

File: BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/Custom Gender Picker/GenderPicker.swift [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+import UIKit
+
+// MARK: - Protocol
+protocol GenderPickerDelegate: class {
+    func view(_view: GenderPicker, needperformAction action: GenderPicker.action )
+}
+
+// MARK: - GenderPicker
+class GenderPicker: UIView {
```
[the remark this answers is not in the document]
Comment (at-thienluong) [reply BaiTap05/TableViewNangCao/ThienLuongQ/TableViewNC/TableViewNangCao/ExPhoneContacts/Custom Gender Picker/GenderPicker.swift]: đổi thành controller hả a

**Pull request merged**